signer/nostr-browser-signer-proxy/src/lib.rs

```diff
@@ -0,0 +1,482 @@
+    state: ProxyState,
+) -> Result<Response<BoxBody<Bytes, Error>>, Error> {
+    match (req.method(), req.uri().path()) {
+        // Serve the HTML proxy page
+        (&Method::GET, "/") => Ok(Response::builder()
+            .header("Content-Type", "text/html")
+            .body(full(HTML))?),
+        // Serve the CSS page style
+        (&Method::GET, "/style.css") => Ok(Response::builder()
+            .header("Content-Type", "text/css")
+            .body(full(CSS))?),
+        // Serve the JS proxy script
+        (&Method::GET, "/proxy.js") => Ok(Response::builder()
+            .header("Content-Type", "application/javascript")
+            .body(full(JS))?),
+        // Browser polls this endpoint to get pending requests
+        (&Method::GET, "/api/pending") => {
+            let mut outgoing = state.outgoing_requests.lock().await;
+
+            let data: Requests<'_> = Requests {
+                requests: &outgoing,
+            };
+            let json: String = serde_json::to_string(&data)?;
+
+            tracing::debug!(
+                "Sending {} pending requests to browser",
+                data.requests.len()
+            );
+
+            // Clear the outgoing requests after sending them
+            outgoing.clear();
+
+            Ok(Response::builder()
+                .header("Content-Type", "application/json")
+                .header("Access-Control-Allow-Origin", "*")
+                .body(full(json))?)
+        }
+        // Get response
+        (&Method::POST, "/api/response") => {
+            // Correctly collect the body bytes from the stream
+            let body_bytes: Bytes = match req.into_body().collect().await {
+                Ok(collected) => collected.to_bytes(),
+                Err(e) => {
+                    tracing::error!("Failed to read body: {e}");
+                    let response = Response::builder()
+                        .status(StatusCode::BAD_REQUEST)
+                        .body(full("Failed to read body"))?;
+                    return Ok(response);
+                }
+            };
+
+            // Handle responses from the browser extension
+            let message: Message = match serde_json::from_slice(&body_bytes) {
+                Ok(json) => json,
+                Err(_) => {
+                    let response = Response::builder()
+                        .status(StatusCode::BAD_REQUEST)
+                        .body(full("Invalid JSON"))?;
+                    return Ok(response);
+                }
+            };
+
+            tracing::debug!("Received response from browser: {:?}", message);
+
+            let mut pending = state.pending_responses.lock().await;
+
+            if let Some(sender) = pending.remove(&message.id) {
+                let result: Result<Value, String> = if let Some(error) = message.error {
+                    Err(error)
+                } else {
+                    Ok(message.result.unwrap_or(Value::Null))
+                };
```

Comment:
Maybe moving this to a function will be better? Something like `message.into_result()` in `Sender::send`.